File: 40-CombinationSumIi/40-CombinationSumIi.cpp
// Last updated: 7/2/2025, 11:55:15 PM
class Solution {
public:
    vector<vector<int>> combinationSum2(vector<int>& candidates, int target) {
        sort(candidates.begin(), candidates.end(), greater<int>());
        
        vector<vector<int>> ans;
        vector<int> lcl;

        help(candidates, 0,target, ans, lcl);
        
        return ans;
        
    }

    void help(vector<int>& arr, int idx, int tar, vector<vector<int>> &ans,
    vector<int> &lcl)
    {
        if(tar == 0)
        {
            ans.push_back(lcl);
            return;
        }

        for(int i = idx ; i < arr.size();i++)
        {
            if(i > idx && arr[i] == arr[i-1]) continue;
            if(tar - arr[i] >= 0)
            {
                lcl.push_back(arr[i]);
                help(arr,i+1,tar-arr[i],ans,lcl);
                lcl.pop_back();
            }
        }
    }
};

File: 40-CombinationSumIi/40-CombinationSumIi.java
// Last updated: 7/2/2025, 11:55:18 PM
class Solution {
    public List<List<Integer>> combinationSum2(int[] candidates, int target) {
        Arrays.sort(candidates);

        List<List<Integer>> res = new ArrayList<>();
        help( res , candidates , target, 0, new ArrayList<>());
        return res;
    }
    public void help(List<List<Integer>> res , int[] arr , int tar , int i, ArrayList<Integer> lcl)
    {
        if(tar <0) return;
        if(tar == 0) {
            res.add(new ArrayList<Integer>(lcl));
            return;
        }
        for( int j = i; j<arr.length ; j++)
        {
            if( i<j && arr[j] == arr[j-1]) continue;

            lcl.add(arr[j]);
            help(res , arr , tar-arr[j],j+1, lcl);
            lcl.remove(lcl.size()-1);
        }
    }
}